test/integration/apiserver/admissionwebhook/admission_test.go

```diff
@@ -481,12 +505,28 @@ func testResourceUpdate(c *testContext) {
 }
 
 func testResourcePatch(c *testContext) {
-	obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+	stubObj, err := getStubObj(c.gvr, c.resource)
 	if err != nil {
 		c.t.Error(err)
 		return
 	}
-	c.admissionHolder.expect(c.gvr, gvk(c.resource.Group, c.resource.Version, c.resource.Kind), v1beta1.Update, obj.GetName(), obj.GetNamespace(), true, true)
+	ns := ""
+	if c.resource.Namespaced {
+		ns = testNamespace
+	}
+	// 1st patch is a create
```

Comment:
this isn't testing what you think:
1. we test create before patch, so the stub object is already created (createOrGetResource would typically have just returned the existing resource here)
2. even if the object did not already exist, the patch request is not sending enough information to create the object (no metadata.name, or any spec fields, etc), so the create would have failed
3. most objects do not allow create on update for merge patches, so even if the object didn't exist and we updated the patch to have sufficient data, most objects would reject this patch request
4. setting `c.admissionHolder.expect` is only half of the equation... you have to call `c.admissionHolder.verify` to actually verify your expectation was met. otherwise, the next expect call (on line 529) wipes out the first expectation (which is the only reason this is passing)

If we want to exercise create-on-update on all resources, the only method that supports that consistently is server-side-apply, so I'd probably defer adding that aspect of the test to that feature

Reply:
So if I had added the verify I could have seen this fail. I'm starting to understand patch a bit better. Thanks for explaining @liggitt!